---
title: "Phylogenetic tree with Treeio"
output:
  html_document:
    df_print: paged
---


```{r include=FALSE}
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#BiocManager::install("treeio")
library(treeio)
#BiocManager::install("ggtree")
library(ggtree)
#install.packages(tidytree)
library(tidytree)
#install.packages("phytools")
library(phytools)
library(ggplot2)
library(dplyr)
#install.packages("ggnewscale")
library(ggnewscale)
#install.packages("rphylopic")
library(rphylopic)
#install.packages("ggimage")
library(ggimage)
library(ggstar)
#BiocManager::install("ggtreeExtra")
library(ggtreeExtra)
```

```{r}
treefile <- "./mrbayes/mito_nuc_mar2022_outFcatus.con.tre"
mrbayes <- read.mrbayes(treefile)
beasttree <- "./beast/monk_mito12_nuc_29_072023_beast-tree_2.tre"
beast <- read.beast(beasttree)
# beast2 <- beast
# beast2@phylo <- bind.tip(beast@phylo, "AY729880", where=52, position= 0, edge.length = 28.81)
# beast_phylo <- as_tibble(beast2@phylo)
# beast_phylo$type <- "original"
# beast_phylo$type[beast_phylo$label == "AY729880"] <- "hack"
# beast_phylo$sd <- NA
# beast_phylo$sd[parent(beast_phylo, 'AY729880')$node] <- 0.05
# beast_tr2 <- as.treedata(beast_phylo)
```

#Rename taxa

```{r}
#write.tree(beast@phylo)
beast_tips <- beast@phylo$tip.label
beast_tips2 <- beast_tips |> gsub(pattern = "_", replacement = " ") |> gsub(pattern = "Phoca groenlandica", replacement = "Pagophilus groenlandicus") |> gsub(pattern = "Phoca fasciata", replacement = "Histriophoca fasciata")
mrbayes_tips <- mrbayes@phylo$tip.label
mrbayes_tips2 <- mrbayes_tips |> gsub(pattern = "_", replacement = " ") |> gsub(pattern = "Phoca groenlandica", replacement = "Pagophilus groenlandicus") |> gsub(pattern = "Phoca fasciata", replacement = "Histriophoca fasciata")
#sort(beast_tips)
#sort(mrbayes_tips)
#t <- data.frame(sort(beast_tips), sort(mrbayes_tips))
t1 <- data.frame(beast_tips, beast_tips2) 
t2 <- data.frame(mrbayes_tips, mrbayes_tips2)
```


```{r}
plotTree(mrbayes@phylo,setEnv=TRUE,offset=0.7)
tiplabels()
clade <- tree_subset(mrbayes, node=1, levels_back=2)
```
```{r}
p <- ggtree(beast@phylo, size=1) + 
  #geom_text(aes(label=node), hjust=-.3) + 
  geom_tiplab(offset = 2, size = 3) 

p +    
  geom_cladelabel(node=56, label="Canidae", color="blue", offset = 250, offset.text = 2, barsize = 0.7) +
  geom_cladelabel(node=54, label="Mustelidae", color="blue", offset = 250, offset.text = 2, barsize = 0.7) +
  geom_cladelabel(node=57, label="Felidae", color="blue", offset = 250, offset.text = 2, barsize = 0.7) +
  geom_cladelabel(node=33, label="Pinipedia", color="red", offset = 250, offset.text = 2, barsize = 0.7) +
  
#  geom_cladelabel(node=4, label="Odobenidae", color="red", offset = 50, align = TRUE) +
#  geom_cladelabel(node=35, label="Otariidae", color="red", offset = 50) +
  geom_cladelabel(node=34, label="Otariidae +\nOdobenidae", color="red", offset = 200, offset.text = 2, barsize = 0.7) + 
  geom_cladelabel(node=37, label="Phocidae", color="red", offset = 200, offset.text = 2, barsize = 0.7) +
  
  geom_cladelabel(node=38, label="Phocinae", color="red", offset = 150, offset.text = 2, barsize = 0.7) +
  geom_cladelabel(node=47, label="Monochinae", color="red", offset = 150, offset.text = 2, barsize = 0.7) +
  
  geom_cladelabel(node=48, label="Lobodontini+\nMirongini", color="red", offset = 100, offset.text = 2, barsize = 0.7) +
  geom_cladelabel(node=52, label="Monochini", color="red", offset = 100, offset.text = 2, barsize = 0.7) +
  theme_tree2() + 
  xlim(0, 350) +
  theme_tree()
```

```{r}
ggtree(beast@phylo) +  geom_tiplab() 
ggtree(mrbayes@phylo) +  geom_tiplab()
#ggtree(beast_tr2, aes(linetype=type)) +  geom_tiplab() +
#  geom_errorbarh(aes(xmin=x-sd, xmax=x+sd, y = y - 0.3), linetype='dashed', height=0.1) +
#  scale_linetype_manual(values = c("hack" = "dashed", 
#                                   "original" = "solid"))
#mergedtree <- merge_tree(mrbayes,beast2)
ggtree(beast) + theme_tree2() + geom_tiplab() + geom_text(aes(x=node, label=round(posterior,2)), position="identity")

#https://garynapier.github.io/ggtree/
# I think this is how you get the first (estimated) date of the Beast tree - i.e. the root. 
# My root year happens to be 0, but if the root date was different, then I suspect the root.edge would be different 
# (i.e. root.edge is not actually in years)
first_date <- beast@phylo$root.edge
# Get the last overall date from the sample names - year after the last underscore
last_date <- 2023
# In ggtree you have to give the last date as a character string in the form of YYYY-MM-DD
last_date_chr <- paste0(as.character(last_date), "-12-31")
# Total n samples
n_samps <- length(beast@phylo$tip.label)


# Expand the x-axis so there's a bit of room at the start and a very large space (80% of the full timeline) 
# to the right for the annotations
x_lim <- c(0, 100)
# Expand the y-axis so there's room at the top and bottom
y_lim <- c(0, n_samps + (n_samps * 0.1))
# Define vertical line colour
v_line_col <- "red"
# Define sequence of year ticks on the x-axis
x_labs_full <- seq(0, 50, by = 5)
# Angle of the labels so they fit in
angle <- 45

p1 <- ggtree(beast) + theme_tree2() + geom_tiplab(offset = 2, size = 3) +   # Tip labels
  # geom_tiplab(align=TRUE, linetype='dashed', linesize=.3, size = 2) +
  # Red confidence bars
  geom_range("height_0.95_HPD", color='red', size=2, alpha=.5) +
  # Text labels for branch posterior scores
  # Note - you can filter the values - here only 0.99 and below are shown.
  # Also can control the position with the vjust arg. 
  geom_text2(aes(label=round(as.numeric(posterior), 3),
                 subset=as.numeric(posterior)< 1,
                 x=branch), vjust=0.5, hjust=1.5, size=2) +
  # Set the y-axis as defined above
  coord_cartesian(ylim = y_lim) +
  # Set the x-axis limits, and year tick labels (don't know why you need 'breaks' as well as 'labels')
  scale_x_continuous(limits = x_lim, breaks = x_labs_full, labels = rev(x_labs_full))+
  # Set the appearance of the year labels, including angle
  theme(axis.text.x = element_text(face = "bold", size = 8, angle = 45, hjust = 1))

p1
```

```{r}
#https://rphylopic.palaeoverse.org/
#https://cran.r-project.org/web/packages/rphylopic/vignettes/b-advanced-ggplot.html
#https://guangchuangyu.github.io/ggtree-book/chapter-ggimage.html#annotate-tips
# Make a data.frame for the PhyloPic names
seal_data <- data.frame(species = beast@phylo$tip.label, uuid = NA)
# Try to get PhyloPic UUIDs for the species names
seal_data$uuid <- sapply(beast@phylo$tip.label,
                               function(x) {
                                 tryCatch(get_uuid(x), error = function(e) NA)
                               })
#seal_data$uuid[seal_data$species == "Arctocephalus_forsteri"] <- seal_data$uuid[seal_data$species == "Arctocephalus_australis"]

seal_data

seal_data$svg <- lapply(seal_data$uuid, function(x) {
                                            tryCatch(get_phylopic(x), error = function(e) NA)})
```


```{r}
p2 <- ggtree(beast) %<+% seal_data %<+% t1 + 
  theme_tree2() + 
  geom_tiplab(aes(label=beast_tips2),offset = 5, size = 3, ) 

p2 + 
  geom_tiplab(aes(image=uuid), geom="phylopic", offset=2, hjust=0, size=.025) +  # Tip labels
  # geom_tiplab(align=TRUE, linetype='dashed', linesize=.3, size = 2) +
  # Red confidence bars
  geom_range("height_0.95_HPD", color='red', size=2, alpha=.5) +
  # Text labels for branch posterior scores
  # Note - you can filter the values - here only 0.99 and below are shown.
  # Also can control the position with the vjust arg. 
  geom_text2(aes(label=round(as.numeric(posterior), 3),
                 subset=as.numeric(posterior)< 1,
                 x=branch), vjust=0.5, hjust=1.5, size=2) +
  # Set the y-axis as defined above
  coord_cartesian(ylim = y_lim) +
  # Set the x-axis limits, and year tick labels (don't know why you need 'breaks' as well as 'labels')
  scale_x_continuous(trans = scales::reverse_trans(),limits = x_lim, breaks = x_labs_full, labels = rev(x_labs_full))+
  # Set the appearance of the year labels, including angle
  theme(axis.text.x = element_text(face = "bold", size = 8, angle = 45, hjust = 1)) +
  #geom_text(aes(label=node), hjust=-.3) + 
  geom_cladelabel(node=56, label="Canidae", color="blue", offset = 250, offset.text = 2, barsize = 0.7, fontsize = 3) +
  geom_cladelabel(node=54, label="Mustelidae", color="blue", offset = 250, offset.text = 2, barsize = 0.7, fontsize = 3) +
  geom_cladelabel(node=57, label="Felidae", color="blue", offset = 250, offset.text = 2, barsize = 0.7, fontsize = 3) +
  geom_cladelabel(node=33, label="Pinipedia", color="red", offset = 250, offset.text = 2, barsize = 0.7, fontsize = 3) +
  
#  geom_cladelabel(node=4, label="Odobenidae", color="red", offset = 50, align = TRUE) +
#  geom_cladelabel(node=35, label="Otariidae", color="red", offset = 50) +
  geom_cladelabel(node=34, label="Otariidae +\nOdobenidae", color="red", offset = 200, offset.text = 2, barsize = 0.7, fontsize = 3) + 
  geom_cladelabel(node=37, label="Phocidae", color="red", offset = 200, offset.text = 2, barsize = 0.7, fontsize = 3) +
  
  geom_cladelabel(node=38, label="Phocinae", color="red", offset = 150, offset.text = 2, barsize = 0.7, fontsize = 3) +
  geom_cladelabel(node=47, label="Monochinae", color="red", offset = 150, offset.text = 2, barsize = 0.7, fontsize = 3) +
  
  geom_cladelabel(node=48, label="Lobodontini+\nMirongini", color="red", offset = 100, offset.text = 2, barsize = 0.7, fontsize = 3) +
  geom_cladelabel(node=52, label="Monochini", color="red", offset = 100, offset.text = 2, barsize = 0.7, fontsize = 3) +
  theme_tree2() + 
  xlim(0, 350) +
  theme_tree()
```
```{r}
p2 + 
  geom_tiplab(aes(image=uuid), geom="phylopic", offset=2, hjust=0, size=.025) +  # Tip labels
  # geom_tiplab(align=TRUE, linetype='dashed', linesize=.3, size = 2) +
  # Red confidence bars
  geom_range("height_0.95_HPD", color='red', size=2, alpha=.5) +
  # Text labels for branch posterior scores
  # Note - you can filter the values - here only 0.99 and below are shown.
  # Also can control the position with the vjust arg. 
  geom_text2(aes(label=round(as.numeric(posterior), 3),
                 subset=as.numeric(posterior)< 1,
                 x=branch), vjust=0.5, hjust=1.5, size=2) +
  # Set the y-axis as defined above
  coord_cartesian(ylim = y_lim) +
  # Set the x-axis limits, and year tick labels (don't know why you need 'breaks' as well as 'labels')
  scale_x_continuous(trans = scales::reverse_trans(),limits = x_lim, breaks = x_labs_full, labels = rev(x_labs_full))+
  # Set the appearance of the year labels, including angle
  theme(axis.text.x = element_text(face = "bold", size = 8, angle = 45, hjust = 1)) +
  #geom_text(aes(label=node), hjust=-.3) + 
  geom_cladelabel(node=56, label="Canidae", color="blue", offset = 75, offset.text = 2, barsize = 0.7, fontsize = 3) +
  geom_cladelabel(node=54, label="Mustelidae", color="blue", offset = 75, offset.text = 2, barsize = 0.7, fontsize = 3) +
  geom_cladelabel(node=57, label="Felidae", color="blue", offset = 75, offset.text = 2, barsize = 0.7, fontsize = 3) +
  geom_cladelabel(node=34, label="Otariidae +\nOdobenidae", color="red", offset = 75, offset.text = 2, barsize = 0.7, fontsize = 3) + 
  geom_cladelabel(node=38, label="Phocinae", color="red", offset = 75, offset.text = 2, barsize = 0.7, fontsize = 3) +
  geom_cladelabel(node=48, label="Lobodontini+\nMirongini", color="red", offset = 75, offset.text = 2, barsize = 0.7, fontsize = 3) +
  geom_cladelabel(node=52, label="Monochini", color="green", offset = 75, offset.text = 2, barsize = 0.7, fontsize = 3) +
  theme_tree2() + 
  xlim(0, 250) +
  theme_tree()
```
##Beast mito + nuc with clade highlighting, phylopics in clades, stars in fossil calibrations, timescale

**Figure 2b** - Time calibrated phylogeny of pinnipeds estimated from full mitogenomes and 16 nuclear genes. Stars represent fossil calibrated nodes, red bars represent 95 % highest Posterior Density - HPD, only posterior probabilities below 1 are shown. Silhouettes were obtained from phylopic.org, with the following authors: *Vulpes lagopes*: T. Michael Keesey, Public Domain; *Mustela nivalis*: Ferran Sayol, CC0; *Felis catus*: Steven Traver, CC0; *Eumetopias jubatus*: Andy Wilson, CC0; *Pusa hispida*: Tracy Heath, CC0; *Hydrurga leptonyx*: Margot Michaux, CC0; and *Neomonachus tropicalis*: Julián Bayona, CC-BY 3.0.

```{r}
clades <- data.frame(
  node = c(56, 54, 57, 34, 38, 48, 52),
  node2 = c(56, 53, 57, 50, 41, 35, 39),
  name = c(
    "Vulpes lagopes",
    "Mustela nivalis",
    "Felis catus",
    "Eumetopias jubatus",
    "Pusa hispida",
    "Hydrurga leptonyx",
    "Neomonachus tropicalis"
    ),
  taxa = c(
    "Canidae",
    "Mustelidae", 
    "Felidae", 
    "Otariidae +\nOdobenidae", 
    "Phocinae", 
    "Lobodontini+\nMirongini", 
    "Monochini"),
  uuid_clade = c(
    "8db421ba-4b42-4bd8-b948-5214ce122f57",
    "20b6096e-2d6d-43c4-acda-fd74f0f91d48",
    "1edbe9ff-c453-47f3-9174-2e1f9c5983de",
    "8d45b927-bc7c-4e98-a7af-7dc69b888f40",
    "157b50ce-b330-4315-b2cd-53a0fa681d10",
    "da451b45-b32c-4982-861d-59d24ccff14f",
    "dcc27b9e-1bdc-4178-ade9-30a3d20c9806")
)

attribution_phylopic <- get_attribution(uuid = clades$uuid_clade)
cal <- data.frame(
  node = c(31, 48, 49, 32, 34, 37, 40),
  name = c("caniformia",
           "Lobodontini + Miroungini",
           "Lobodontini",
           "Musteloidea + Pinnipedia",
           "Otariodea",
           "Phocidae",
           "Phocina"),
  shape = "*"
)

x_lim <- c(0, 120)

p3 <- p2 %<+% cal + 
    #geom_tiplab(aes(image=uuid), geom="phylopic", offset=2, hjust=0, size=.025) +  # Tip labels
    # geom_tiplab(align=TRUE, linetype='dashed', linesize=.3, size = 2) +
    # Red confidence bars
    geom_range("height_0.95_HPD", color='red', size=2, alpha=.5) +
    # Text labels for branch posterior scores
    # Note - you can filter the values - here only 0.99 and below are shown.
    # Also can control the position with the vjust arg. 
    geom_text2(aes(label=round(as.numeric(posterior), 3),
                   subset=as.numeric(posterior)< 1,
                   x=branch), vjust=0.5, hjust=1.5, size=2) +
    # Set the y-axis as defined above
    coord_cartesian(ylim = y_lim) +
    # Set the x-axis limits, and year tick labels (don't know why you need 'breaks' as well as 'labels')
    scale_x_continuous(trans = scales::reverse_trans(),limits = x_lim, breaks = x_labs_full, labels = rev(x_labs_full))+
    # Set the appearance of the year labels, including angle
    theme(axis.text.x = element_text(face = "bold", size = 8, angle = 45, hjust = 1)) +
    #geom_text(aes(label=node), hjust=-.3) + 
#    geom_fruit(data=clades, mapping = aes(node = node, label= taxa, uuid = uuid_clade, color = taxa), geom = geom_phylopic, fontsize = 5, align = TRUE, show.legend = FALSE) +
    geom_cladelab(data = clades, mapping = aes(node = node, label= taxa, image = uuid_clade, color = taxa), geom = "phylopic", fontsize = 5, align = FALSE, show.legend = FALSE, offset = 50, offset.text = 5, barsize = 2) +
  #scale_color_brewer(type= 'qual', palette = "Dark2", breaks=c('Felidae', 'Canidae', 'Mustelidae', 'Otariidae +\nOdobenidae', 'Phocinae', 'Lobodontini+\nMirongini', 'Monochini')) +
  scale_color_manual(values = c("#1B9E77", "#D95F02", "#7570B3", "#E7298A", "#66A61E", "#E6AB02", "#A6761D"), breaks = c('Felidae', 'Canidae', 'Mustelidae', 'Otariidae +\nOdobenidae', 'Phocinae', 'Lobodontini+\nMirongini', 'Monochini')) +
    theme_tree2() + 
#    xlim(0, 120) +
#    theme_tree()
  geom_nodelab(mapping = aes(node=node, text=shape)) +
  geom_text2(aes(label=shape), vjust=0.5, hjust=2.5, size=4) +
# geom_nodepoint(mapping = aes(node = node, shape=shape), show.legend = FALSE) +
  coord_cartesian(ylim = y_lim) +
  # Set the x-axis limits, and year tick labels (don't know why you need 'breaks' as well as 'labels')
  scale_x_continuous(limits = x_lim, breaks = x_labs_full, labels = rev(x_labs_full))+
  # Set the appearance of the year labels, including angle
  theme(axis.text.x = element_text(face = "bold", size = 8, angle = 45, hjust = 1))

p3

ggsave(filename = "Figure 2b Beast mito plus nuc.pdf", plot= p3)
```

##Topology comparison Beast vs. MrBayes

```{r}
p1 <- ggtree(mrbayes@phylo,size=1)
p1 <- flip(p1, 22, 23) %>% flip(39, 38) %>% flip(7, 8) %>% flip(2, 3)
p2 <- ggtree(beast@phylo,size=2)

d1 <- p1$data
d2 <- p2$data

## reverse x-axis and 
## set offset to make the tree in the right hand side of the first tree
d2$x <- (max(d2$x) - d2$x)/30 + max(d1$x) + 1

pp <- p1 + geom_tree(data=d2) +      
  ggnewscale::new_scale_fill()
# +
#    geom_hilight(
#          data = d2, 
#          mapping = aes( 
#             subset = node %in% c(1, 2),
#             node=node,
#            fill=as.factor(node))
#    ) +
#   labs(fill = "Neomonachus" ) 

dd <- bind_rows(d1, d2) %>% 
  filter(!is.na(label))

pp + geom_line(aes(x, y, group=label), data=dd, color='firebrick') +
    geom_tiplab(align=1) + geom_nodelab() +
    geom_tiplab(data = d2, hjust=1) 

pp + geom_nodelab()
```

##Mr Bayes Mitochondrial Tree

**Figure 2a** - Bayesian phylogeny of pinnipedia using complete mitochondrial genes, excluding ND6. Phylogeny is rooted with Felidae, Canidae and Mustelidae species. Node labels indicate posterior probabilities below 1. Vertical coloured bars indicate clades.

```{r}
#treefile2 <- "./arvoreMrBayesfinalmito_concat_mar2022.con (1).tre"
treefile2 <- "./mrbayes/mito_concat_mar2022.con.tre"
#treefile3 <- "./mrbayes/mitogenomes_mar2022.con.tre"
mb_mito <- read.mrbayes(treefile2)

#Replacing the tip labels with names with spaces and updated taxonomy
mb_tips <- mb_mito@phylo$tip.label
mb_tips2 <- mb_tips |> gsub(pattern = "_", replacement = " ") |> gsub(pattern = "Phoca groenlandica", replacement = "Pagophilus groenlandicus") |> gsub(pattern = "Phoca fasciata", replacement = "Histriophoca fasciata") |> gsub(pattern = " concolor", replacement = "")
t3 <- data.frame(mb_tips, mb_tips2)

ggtree(mb_mito) + geom_text(aes(label=node), hjust=-.3) + geom_tiplab(align = TRUE, offset = 0.1) + xlim(NA, 1)
#mb_mito3 <- read.mrbayes(treefile3)
mb_mito@phylo <- reroot(mb_mito@phylo, 38)
ggtree(mb_mito) + geom_text(aes(label=node), hjust=-.3) + geom_tiplab(align = TRUE)

#Uses t3 to replace "_" with " " and update species names in tip labels

p_mito <- ggtree(mb_mito, size = 1) %<+% t3 + theme_tree() + geom_tiplab(aes(label = mb_tips2), align = TRUE, offset = 0.01, size = 3)
p_mito <- flip(p_mito, 57, 56)


p_mito_publish <- p_mito + xlim(NA, 1.5) + ylim(-1, NA) + geom_treescale(y = -1, fontsize = 2, linesize = 1, width = 0.1) + theme_tree() + geom_text2(aes(label=round(as.numeric(prob), 2),
                                                 subset=as.numeric(prob)< 1, 
                                                 x=branch), vjust=-1.0, size=2) + 
  geom_cladelab(data = clades, mapping = aes(node = node2, label= taxa, image = uuid_clade, color = taxa), geom = "phylopic", fontsize = 8, align = TRUE, show.legend = FALSE, offset = 0.6, offset.text = 0.1, barsize = 2) +
  scale_color_manual(values = c("#1B9E77", "#D95F02", "#7570B3", "#E7298A", "#66A61E", "#E6AB02", "#A6761D"), breaks = c('Felidae', 'Canidae', 'Mustelidae', 'Otariidae +\nOdobenidae', 'Phocinae', 'Lobodontini+\nMirongini', 'Monochini'))
  #scale_color_brewer(type= 'qual', palette = "Dark2", breaks=c('Felidae', 'Canidae', 'Mustelidae', 'Otariidae +\nOdobenidae', 'Phocinae', 'Lobodontini+\nMirongini', 'Monochini')) 
#ggtree(mb_mito3) + theme_tree() + geom_tiplab(align = TRUE, size = 3)

p_mito_publish

ggsave(filename = "Figure 2a MrBayes mitogenomes.pdf", p_mito_publish)
```

# Choosing custom colors to use in palette so that they match in different plots

```{r}
library(RColorBrewer)
my_palette = c(brewer.pal(5, "Set1")[c(1,3,4,5)], brewer.pal(5, "Pastel1")[c(2,5,1,3)])
#grid::grid.raster(my_palette, int=F)

scale_colour_discrete = function(...) scale_colour_manual(..., values = palette())

dsamp <- diamonds[sample(nrow(diamonds), 1000), ]
(p <- qplot(carat, price, data = dsamp, colour = clarity)) # default palette

palette(my_palette)
p # custom colors

display.brewer.all()
display.brewer.pal(7, "Dark2")
brewer.pal(7, "Dark2")
```

